File: src/main/java/alfheim/common/spell/sound/SpellEcho.java
package alfheim.common.spell.sound;

import java.util.List;

import alexsocol.asjlib.math.Vector3;
import alfheim.AlfheimCore;
import alfheim.api.entity.EnumRace;
import alfheim.api.spell.SpellBase;
import alfheim.client.render.world.SpellEffectHandlerClient.Spells;
import alfheim.common.core.handler.SpellEffectHandler;
import alfheim.common.network.MessageParticles;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.monster.IMob;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;

public class SpellEcho extends SpellBase {

	public SpellEcho() {
		super("echo", EnumRace.POOKA, 4000, 1500, 5);
	}

	@Override
	public SpellCastResult performCast(EntityLivingBase caster) {
		SpellCastResult result = checkCast(caster);
		if (result != SpellCastResult.OK) return result;
		
		List<Entity> list = caster.worldObj.getEntitiesWithinAABB(Entity.class, caster.boundingBox.expand(16, 16, 16));
		for (Entity e: list) {
			if (Vector3.entityDistance(e, caster) > 16) continue;
			if (e instanceof EntityItem) SpellEffectHandler.sendPacket(Spells.ECHO_ITEM, e);
			else if (e instanceof IMob) SpellEffectHandler.sendPacket(Spells.ECHO_MOB, e);
			else if (e instanceof EntityPlayer) SpellEffectHandler.sendPacket(Spells.ECHO_PLAYER, e);
			else if (e instanceof EntityLivingBase) SpellEffectHandler.sendPacket(Spells.ECHO_ENTITY, e);
		}
		
		if (caster instanceof EntityPlayerMP) AlfheimCore.network.sendTo(new MessageParticles(Spells.ECHO.ordinal(), caster.posX, caster.posY, caster.posZ), (EntityPlayerMP) caster);
		return result;
	}
}

File: src/main/java/alfheim/common/spell/sound/SpellEcho.kt
package alfheim.common.spell.sound

import alexsocol.asjlib.expand
import alexsocol.asjlib.math.Vector3
import alexsocol.asjlib.security.InteractionSecurity
import alfheim.AlfheimCore
import alfheim.api.entity.EnumRace
import alfheim.api.spell.SpellBase
import alfheim.client.render.world.VisualEffectHandlerClient.VisualEffects
import alfheim.common.core.handler.VisualEffectHandler
import alfheim.common.network.MessageVisualEffect
import net.minecraft.entity.*
import net.minecraft.entity.item.EntityItem
import net.minecraft.entity.monster.IMob
import net.minecraft.entity.player.*

object SpellEcho: SpellBase("echo", EnumRace.POOKA, 4000, 1500, 5) {
	
	override val usableParams: Array<Any>
		get() = arrayOf(radius)
	
	override fun performCast(caster: EntityLivingBase): SpellCastResult {
		val result = checkCast(caster)
		if (result != SpellCastResult.OK) return result
		
		val list = caster.worldObj.getEntitiesWithinAABB(Entity::class.java, caster.boundingBox.expand(radius)) as List<Entity>
		for (e in list) {
			if (Vector3.entityDistance(e, caster) > radius * 2) continue
			
			if (!InteractionSecurity.canDoSomethingWithEntity(caster, e)) continue
			
			when (e) {
				is EntityItem       -> VisualEffectHandler.sendPacket(VisualEffects.ECHO_ITEM, e)
				is IMob             -> VisualEffectHandler.sendPacket(VisualEffects.ECHO_MOB, e)
				is EntityPlayer     -> VisualEffectHandler.sendPacket(VisualEffects.ECHO_PLAYER, e)
				is EntityLivingBase -> VisualEffectHandler.sendPacket(VisualEffects.ECHO_ENTITY, e)
			}
		}
		
		if (caster is EntityPlayerMP) AlfheimCore.network.sendTo(MessageVisualEffect(VisualEffects.ECHO.ordinal, caster.posX, caster.posY, caster.posZ), caster)
		return result
	}
}